ext/sqlite3/database.c

```diff
@@ -12,6 +12,39 @@
 
 VALUE cSqlite3Database;
 
+static void
+close_or_discard_db(sqlite3RubyPtr ctx)
+{
+    if (ctx->db) {
+        if (ctx->owner == getpid()) {
+            // Ordinary close.
+            sqlite3_close_v2(ctx->db);
+        } else {
+            // This is an open connection carried across a fork().
+            // "Discard" it. See adr/2024-09-fork-safety.md
+            sqlite3_file *sfile;
+            int status;
+
+            rb_warning("An open sqlite database connection was inherited from a forked process and "
```

Comment:
I'm starting to wonder? This doesn't apply to read only connections right? Might be nice to not leak for them?

Reply:
> This doesn't apply to read only connections right

My reading of §2.6 of https://www.sqlite.org/howtocorrupt.html does not exclude readonly connections. I see it mentions "locking problems" but I don't think that sentence is intended to be a canonical list of conditions. I'm choosing to pay attention to:

> SQLite is not designed to support that kind of behavior. Any database connection that is used in a child process must be opened in the child process.

which is unconditional.

I'm choosing to be extremely conservative here because I think the negative impact of data loss justifies it. We can always loosen the "discard" conditions if we learn more.

Reply:
> I'm choosing to be extremely conservative here because I think the negative impact of data loss justifies it.

Yeah, but if the connection is read only, I don't see how there could possibly be data loss. I have some planned used for read only sqlite as some static data store, and not leaking memory on every refork would be preferable.

Reply:
Keep in mind I don't fully understand the nature of how sqlite implements locking, but reading through this entry in particular seems relevant:

> suppose a multi-thread process has two or more threads with separate SQLite database connections to the same database file. Then a third thread comes along and wants to read something out of that same database file on its own, without using the SQLite library. The third thread does an open(), a read() and then a close(). One would think this would be harmless. But the close() system call caused the locks held on the database by all the other threads to be dropped. Those other threads have no way of knowing that their locks have just been trashed

The implication here is that another thread in the process could be writing, and the presence of a read-only connection is still going to interact with locks on the database and could screw up the writer(s).

I suppose we could implement tracking of which `Database` objects are pointed at the same database file and which are writers and have additional logic in the `close_or_discard_db` function ... but that doesn't feel justified ... 🤷 why not just open the readonly connection in the child?

Reply:
Like, I'm open to adding more specific conditions to the check we're doing ... but not in this PR. I want to land this and cut a release. Then we can take the time to make it work more optimally for some of these cases. Does that sound fair?

Reply:
Yeah, of course.

Reply:
> Like, I'm open to adding more specific conditions to the check we're doing ... but not in this PR.

👍 

Since you _can_ open a database read-only, it seems fine to do more specific checks. I agree with shipping this first though.